hw/ip/otbn/util/insn_yaml.py

```diff
@@ -1101,6 +1319,64 @@ def get_masks(self) -> Tuple[int, int]:
         assert (m0 | m1) == all_bits
         return (m0, m1)
 
+    def get_ones_mask(self) -> int:
+        '''Return the mask of fixed bits that are set
+
+        For literal values of x (unused bits in the encoding), we'll prefer
+        '0'.
+
+        '''
+        m0, m1 = self.get_masks()
+        return m1 & ~m0
+
+    def assemble(self, op_to_idx: Dict[str, int]) -> int:
+        '''Assemble an instruction
+
+        op_to_idx should map each operand in the encoding to some integer
+        index, which should be small enough to fit in the width of the
+        operand's type and should be representable after any shift. Will raise
+        a ValueError if not.
+
+        '''
+        val = self.get_ones_mask()
+        for field_name, field in self.fields.items():
+            if not isinstance(field.value, str):
+                # We've done this field already (in get_ones_mask)
+                continue
+
+            # Try to get the operand value for the field. If this is an
+            # optional operand, we might not have one, and just encode zero.
+            field_val = op_to_idx.get(field.value, 0)
+
+            # Are there any low bits that shouldn't be there?
+            shift_mask = (1 << field.scheme_field.shift) - 1
+            if field_val & shift_mask:
+                raise ValueError("operand field {} has a shift of {}, "
+                                 "so can't represent the value {:#x}."
+                                 .format(field.value,
+                                         field.scheme_field.shift,
+                                         field_val))
+
+            shifted = field_val >> field.scheme_field.shift
+
+            # Is the number too big? At the moment, we are assuming immediates
+            # are unsigned (because the OTBN instructions all have unsigned
```

Comment:
```suggestion
            # are unsigned (because the BN instructions all have unsigned
```

RV32I has signed immediates generally (e.g. ADDI).

Reply:
Good point, I'll take that change. The logic isn't strictly correct here, because the code doesn't understand that `addi` or (more interestingly) `jal` take signed immediates. However, it also doesn't know how to turn such immediates from a string into a number, so it doesn't matter: we just pass the text through to the underlying assembler, which should hopefully get it right.

I suspect we might have to come back to this and be a bit more careful in the future, but I thought I'd do the simple thing for now and worry about it if we need to.